src/rust/sqs-executor/src/redis_cache.rs

```diff
@@ -0,0 +1,281 @@
+
+
+#[derive(thiserror::Error, Debug)]
+pub enum RedisCacheError {
+    #[error("RedisError: {0}")]
+    RedisError(#[from] RedisError),
+    #[error("Cache Timeout: {0}")]
+    Timeout(#[from] Elapsed),
+    #[error("JoinError: {0}")]
+    JoinError(#[from] tokio::task::JoinError),
+}
+
+impl CheckedError for RedisCacheError {
+    fn error_type(&self) -> Recoverable {
+        match self {
+            RedisCacheError::RedisError(RedisError::EmptySlice) => Recoverable::Persistent,
+            _ => Recoverable::Transient,
+        }
+    }
+}
+
+// todo: We should add a circuit breaker to this and not hit the backing redis-store if it closes
+
+#[derive(Clone)]
+pub struct RedisCache {
+    address: String,
+    connection_pool: ConnectionPool,
+    metric_reporter: MetricReporter<Stdout>,
+    lru_cache: Arc<Mutex<lru_cache::LruCache<Vec<u8>, ()>>>,
+}
+
+impl RedisCache {
+    pub async fn new(
+        address: String,
+        metric_reporter: MetricReporter<Stdout>,
+    ) -> Result<Self, RedisError> {
+        let connection_pool =
+            ConnectionPool::create(address.clone(), None, num_cpus::get() * 10).await?;
+
+        Ok(Self {
+            connection_pool,
+            address,
+            metric_reporter,
+            lru_cache: Arc::new(Mutex::new(lru_cache::LruCache::new(100_000)))
+        })
+    }
+}
+
+impl RedisCache {
+    #[tracing::instrument(skip(self, cacheable))]
+    async fn _get<CA>(&mut self, cacheable: CA) -> Result<CacheResponse, RedisCacheError>
+    where
+        CA: Cacheable + Send + Sync + 'static,
+    {
+        let identity_bytes = cacheable.identity();
+        {
+            let mut lru_cache = self.lru_cache.lock().unwrap();
+            if lru_cache.contains_key(&identity_bytes) {
+                return Ok(CacheResponse::Hit);
+            }
+        }
+
+        let identity = hex::encode(&identity_bytes);
+        //
+        let mut client = self.connection_pool.get().await;
+
+        let res = tokio::time::timeout(Duration::from_millis(200), client.exists(&identity)).await;
+
+        let res = match res {
+            Ok(res) => res,
+            Err(e) => {
+                warn!(
+                    errors=e.to_string().as_str(),
+                    "Cache lookup failed with"
+                );
+                return Ok(CacheResponse::Miss);
+            }
+        };
+
+        match res {
+            Ok(true) => {
+                let mut lru_cache = self.lru_cache.lock().unwrap();
+                lru_cache.insert(identity_bytes, ());
+                Ok(CacheResponse::Hit)
+            },
+            Ok(false) => Ok(CacheResponse::Miss),
+            Err(e) => {
+                warn!(error=e.to_string().as_str(), "Cache lookup failed with");
+                Ok(CacheResponse::Miss)
+            }
+        }
+    }
+
+    #[tracing::instrument(skip(self, identity))]
+    async fn _store(&mut self, identity: Vec<u8>) -> Result<(), RedisCacheError> {
```

Comment:
just an idea, not a blocker: wrapping or NewType'ing identity type instead of the sort of odd vec<u8>